extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/LazyObjectSummariesIterator.java

```diff
@@ -0,0 +1,162 @@
+ *
+ * Walks a set of prefixes, returning all objects underneath them except for directory placeholders.
+ */
+public class LazyObjectSummariesIterator implements Iterator<S3ObjectSummary>
+{
+  private final ServerSideEncryptingAmazonS3 s3Client;
+  private final Iterator<URI> prefixesIterator;
+  private final int maxListingLength;
+
+  private ListObjectsV2Request request;
+  private ListObjectsV2Result result;
+  private Iterator<S3ObjectSummary> objectSummaryIterator;
+  private S3ObjectSummary currentObjectSummary;
+
+  LazyObjectSummariesIterator(
+      final ServerSideEncryptingAmazonS3 s3Client,
+      final Iterable<URI> prefixes,
+      final int maxListingLength
+  )
+  {
+    this.s3Client = s3Client;
+    this.prefixesIterator = prefixes.iterator();
+    this.maxListingLength = maxListingLength;
+
+    prepareNextRequest();
+    fetchNextBatch();
+    advanceObjectSummary();
+  }
+
+  private void prepareNextRequest()
+  {
+    final URI currentUri = prefixesIterator.next();
+    final String currentBucket = currentUri.getAuthority();
+    final String currentPrefix = S3Utils.extractS3Key(currentUri);
+
+    request = new ListObjectsV2Request()
+        .withBucketName(currentBucket)
+        .withPrefix(currentPrefix)
+        .withMaxKeys(maxListingLength);
+  }
+
+  private void fetchNextBatch()
+  {
+    try {
+      result = S3Utils.retryS3Operation(() -> s3Client.listObjectsV2(request));
+      request.setContinuationToken(result.getNextContinuationToken());
+      objectSummaryIterator = result.getObjectSummaries().iterator();
+    }
+    catch (Exception e) {
+      throw new RE(
+          e,
+          "Failed to get object summaries from S3 bucket[%s], prefix[%s]",
+          request.getBucketName(),
+          request.getPrefix()
+      );
+    }
+  }
+
+  /**
+   * Advance objectSummaryIterator to the next non-placeholder, updating "currentObjectSummary".
+   */
+  private void advanceObjectSummary()
+  {
+    while (objectSummaryIterator.hasNext() || result.isTruncated() || prefixesIterator.hasNext()) {
+      while (objectSummaryIterator.hasNext()) {
+        currentObjectSummary = objectSummaryIterator.next();
+
+        if (!isDirectoryPlaceholder(currentObjectSummary)) {
+          return;
+        }
+      }
+
+      // Exhausted "objectSummaryIterator" without finding a non-placeholder.
+      if (result.isTruncated()) {
+        fetchNextBatch();
+      } else if (prefixesIterator.hasNext()) {
+        prepareNextRequest();
+        fetchNextBatch();
+      }
+    }
+
+    // Truly nothing left to read.
+    currentObjectSummary = null;
+  }
+
+  @Override
+  public boolean hasNext()
+  {
+    return currentObjectSummary != null;
+  }
+
+  @Override
+  public S3ObjectSummary next()
+  {
+    if (currentObjectSummary == null) {
+      throw new NoSuchElementException();
+    }
+
+    final S3ObjectSummary retVal = currentObjectSummary;
+    advanceObjectSummary();
+    return retVal;
+  }
```

Comment:
I'll move it, because I agree with this particular point, but IMO style nits should either be encoded into an automated style checker or written down somewhere centrally agreed-upon. Otherwise they can create friction in the review process. Maybe we can come up with a nice way as a community to handle this.